corehq/apps/locations/views.py

```diff
@@ -102,42 +100,121 @@ def page_context(self):
-            if cleaned_code != loctype['code']:
-                err = _(
-                    'Location type code "{code}" is invalid. No spaces or special characters are allowed. '
-                    'It has been replaced with "{new_code}".'
-                )
-                messages.warning(request, err.format(code=loctype['code'], new_code=cleaned_code))
-                loctype['code'] = cleaned_code
-            return LocationType(**loctype)
-
-        #TODO add server-side input validation here (currently validated on client)
+        def mk_loctype(name, code, allowed_parents, administrative,
+                       shares_cases, view_descendants):
+            if allowed_parents and allowed_parents[0]:
+                parent = sql_loc_types[allowed_parents[0]]
+            else:
+                parent = None
 
-        self.domain_object.location_types = [mk_loctype(l) for l in payload['loc_types']]
+            cleaned_code = unicode_slug(code)
+            if cleaned_code != code:
+                err = _('Location type code "{code}" is invalid. No spaces or '
+                        'special characters are allowed. It has been replaced '
+                        'with "{new_code}".')
+                messages.warning(request,
+                                 err.format(code=code, new_code=cleaned_code))
 
-        self.domain_object.save()
+            try:
+                loc_type = LocationType.objects.get(domain=self.domain, name=name)
+            except LocationType.DoesNotExist:
+                loc_type = LocationType(domain=self.domain, name=name)
+            loc_type.code = cleaned_code
+            loc_type.administrative = administrative
+            loc_type.parent_type = parent
+            loc_type.shares_cases = shares_cases
+            loc_type.view_descendants = view_descendants
+            loc_type.save()
+            sql_loc_types[name] = loc_type
+
+        loc_types = payload['loc_types']
+        for loc_type in loc_types:
+            for prop in ['name', 'code', 'allowed_parents', 'administrative',
+                         'shares_cases', 'view_descendants']:
+                assert prop in loc_type, "Missing a location type property!"
+
+        heirarchy = self.get_heirarchy(loc_types)
+        for loc_type in heirarchy:
+            mk_loctype(**loc_type)
 
         return self.get(request, *args, **kwargs)
 
+    # This is largely copy-pasted from the LocationTypeManager
+    def get_heirarchy(self, loc_types):
+        """
+        Return loc types in order from parents to
+        """
+        lt_dict = {lt['name']: lt for lt in loc_types}
+
+        # Make sure there are no cycles
+        for loc_type in loc_types:
+            visited = set()
+
+            def step(lt):
+                assert lt['name'] not in visited, \
+                    "There's a loc type cycle, we need to prohibit that"
+                visited.add(lt['name'])
+                parents = lt['allowed_parents']
```

Comment:
seems like this could benefit from an assert that `allowed_parents` is at most one and then you can modify the `mk_loctype` accordingly